week7/answers.sql

```diff
@@ -1,3 +1,212 @@
+  inventory_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+  character_id INT UNSIGNED,
+  item_id INT UNSIGNED,
+  FOREIGN KEY (character_id) REFERENCES characters(character_id),
+  FOREIGN KEY (item_id) REFERENCES items(item_id) 
+);
+
+CREATE TABLE equipped (
+  equipped_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+  character_id INT UNSIGNED,
+  item_id INT UNSIGNED,
+  FOREIGN KEY (character_id) REFERENCES characters(character_id),
+  FOREIGN KEY (item_id) REFERENCES items(item_id)
+);
+
+-- Create Views
+CREATE VIEW character_items AS 
+SELECT 
+  c.character_id,
+  c.name AS character_name,
+  i.name AS item_name,
+  i.armor,
+  i.damage
+FROM characters c
+INNER JOIN (
+  SELECT character_id, item_id FROM inventory
+  UNION
+  SELECT character_id, item_id FROM equipped)
+  AS ce ON c.character_id = ce.character_id
+INNER JOIN items i ON ce.item_id = i.item_id
+ORDER BY c.character_id, i.name;
+
+CREATE VIEW team_items AS
+SELECT 
+  t.team_id,
+  t.name AS team_name,
+  i.name AS item_name,
+  i.armor,
+  i.damage
+FROM teams t 
+INNER JOIN team_members tm ON t.team_id = tm.team_id
+INNER JOIN (
+  SELECT character_id, item_id FROM inventory
+  UNION 
+  SELECT character_id, item_id FROM equipped)
+  AS te ON tm.character_id = te.character_id
+INNER JOIN items i ON te.item_id = i.item_id
+ORDER BY t.team_id, i.name;
+
+-- Create function
+DELIMITER ;;
+
+CREATE FUNCTION armor_total(p_character_id INT UNSIGNED) RETURNS INT
+DETERMINISTIC
+BEGIN
+  DECLARE total_armor INT;
+
+  SELECT COALESCE(SUM(armor), 0)
+  INTO total_armor
+  FROM character_stats
+  WHERE character_id = p_character_id;
+
+  SELECT COALESCE(SUM(items.armor), 0)
+  INTO total_armor
+  FROM equipped
+  INNER JOIN items ON equipped.item_id = items.item_id
+  WHERE equipped.character_id = p_character_id;
+
+  RETURN total_armor;
+END ;;
+
+DELIMITER ;
+
+DELIMITER ;;
+
+CREATE PROCEDURE attack(
+  p_id_of_character_being_attacked INT UNSIGNED,
+  p_id_of_equipped_item_used_for_attack INT UNSIGNED
+)
+BEGIN
+  DECLARE character_armor INT;
+  DECLARE item_damage INT;
+  DECLARE damage_dealt INT;
+
+  SET character_armor = armor_total(p_id_of_character_being_attacked);
+
+  SELECT damage INTO item_damage
+  FROM equipped
+  JOIN items ON equipped.item_id = items.item_id
+  WHERE equipped.equipped_id = p_id_of_equipped_item_used_for_attack;
+
+  SET damage_dealt = GREATEST(item_damage - character_armor, 0);
+
+  UPDATE character_stats
+  SET health = GREATEST(health - damage_dealt, 0)
+  WHERE character_id = p_id_of_character_being_attacked;
+
+  IF health <= 0 THEN
+    DELETE FROM characters WHERE character_id = p_id_of_character_being_attacked;
+    DELETE FROM team_members WHERE character_id = p_id_of_character_being_attacked;
+    DELETE FROM winners WHERE character_id = p_id_of_character_being_attacked;
+  END IF;
+END ;;
+
+DELIMITER ;
+
+DELIMITER ;;
+
+CREATE PROCEDURE equip(p_inventory_id INT UNSIGNED)
+BEGIN
+  DECLARE item_id INT;
```

Comment:
I think this *might* have worked if lines 164-168 were removed (I think "item_id" is confusing the interpreter due to ambiguity). This can be performed with two statements (not counting the DELETE FROM) and without a nested SELECT. As is, it isn't passing the automated test.

Reply:
Same comments for unequip.